File: engine.py
"""The replay engine.

The engine streams a trace through a policy in timestamp order and records,
for each request, whether it was a cache HIT or MISS, the match quality, and
the latency outcome. On a MISS the response is *inserted* into the cache so
later requests can hit it — exactly how a real cache warms up.

Latency model
-------------
- A MISS costs the request's recorded ``cost_ms`` (the origin must serve it).
- A HIT costs a fixed, configurable ``hit_cost_ms`` (the cache lookup itself).
- *Avoided latency* for a hit is ``cost_ms - hit_cost_ms`` (never negative).

This keeps the model honest: caches are not free, and a hit that saves almost
nothing is reported as such.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable

from .normalize import derive
from .strategy import Strategy
from .trace import Request


@dataclass
class Event:
    """The outcome of replaying one request."""

    id: str
    route: str
    hit: bool
    quality: float
    kind: str
    served_ms: float
    avoided_ms: float


@dataclass
class ReplayResult:
    """Aggregate outcome of replaying a whole trace through one strategy."""

    strategy: str
    events: list[Event] = field(default_factory=list)

    @property
    def total(self) -> int:
        return len(self.events)

    @property
    def hits(self) -> int:
        return sum(1 for e in self.events if e.hit)

    @property
    def hit_rate(self) -> float:
        return round(self.hits / self.total, 4) if self.total else 0.0

    @property
    def avoided_ms(self) -> float:
        return round(sum(e.avoided_ms for e in self.events), 3)

    @property
    def served_ms(self) -> float:
        return round(sum(e.served_ms for e in self.events), 3)

    @property
    def mean_quality(self) -> float:
        hit_events = [e.quality for e in self.events if e.hit]
        if not hit_events:
            return 0.0
        return round(sum(hit_events) / len(hit_events), 4)

    def per_route(self) -> dict[str, dict[str, float]]:
        acc: dict[str, dict[str, float]] = {}
        for e in self.events:
            bucket = acc.setdefault(e.route, {"total": 0, "hits": 0, "avoided_ms": 0.0})
            bucket["total"] += 1
            bucket["hits"] += int(e.hit)
            bucket["avoided_ms"] += e.avoided_ms
        for bucket in acc.values():
            total = bucket["total"] or 1
            bucket["hit_rate"] = round(bucket["hits"] / total, 4)
            bucket["avoided_ms"] = round(bucket["avoided_ms"], 3)
        return dict(sorted(acc.items()))


def replay(
    requests: Iterable[Request],
    strategy: Strategy,
    *,
    hit_cost_ms: float = 5.0,
    drop_stopwords: bool = True,
) -> ReplayResult:
    """Replay *requests* through *strategy* and return a :class:`ReplayResult`.

    Requests are processed in the order given; sort upstream if you need
    strict timestamp ordering. Each miss is inserted so the cache warms up
    naturally over the trace.
    """

    ordered = sorted(requests, key=lambda r: (r.ts, r.id))
    result = ReplayResult(strategy=strategy.name)

    for req in ordered:
        signals = derive(req.prompt, drop_stopwords=drop_stopwords)
        match = strategy.lookup(signals, req.route)
        if match is not None:
            served = min(hit_cost_ms, req.cost_ms)
            avoided = max(req.cost_ms - served, 0.0)
            result.events.append(
                Event(
                    id=req.id,
                    route=req.route,
                    hit=True,
                    quality=match.quality,
                    kind=match.kind,
                    served_ms=round(served, 3),
                    avoided_ms=round(avoided, 3),
                )
            )
        else:
            strategy.insert(req.id, signals, req.route)
            result.events.append(
                Event(
                    id=req.id,
                    route=req.route,
                    hit=False,
                    quality=0.0,
                    kind="miss",
                    served_ms=round(req.cost_ms, 3),
                    avoided_ms=0.0,
                )
            )

    return result

File: engine.go
package craft

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"sort"
	"strings"
)

// Request is one recorded request from a trace.
type Request struct {
	ID       string  `json:"id"`
	Route    string  `json:"route"`
	Prompt   string  `json:"prompt"`
	CostMS   float64 `json:"cost_ms"`
	TS       float64 `json:"ts"`
	TokensIn int     `json:"tokens_in"`
}

// LoadTrace parses a JSONL request trace from r. Blank lines and lines that
// begin with '#' are ignored so fixtures can carry comments.
func LoadTrace(r io.Reader) ([]Request, error) {
	var out []Request
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		var req Request
		if err := json.Unmarshal([]byte(line), &req); err != nil {
			return nil, fmt.Errorf("line %d: %w", lineNo, err)
		}
		if req.Route == "" || req.Prompt == "" {
			return nil, fmt.Errorf("line %d: route and prompt are required", lineNo)
		}
		if req.ID == "" {
			req.ID = fmt.Sprintf("req-%06d", lineNo)
		}
		out = append(out, req)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return out, nil
}

// Result aggregates a replay of one strategy over a trace.
type Result struct {
	Strategy  string  `json:"strategy"`
	Total     int     `json:"requests"`
	Hits      int     `json:"hits"`
	HitRate   float64 `json:"hit_rate"`
	AvoidedMS float64 `json:"avoided_ms"`
	ServedMS  float64 `json:"served_ms"`
	MeanQual  float64 `json:"mean_quality"`
}

// Replay streams requests through a strategy, inserting misses so the cache
// warms over the trace. It returns aggregate metrics.
func Replay(reqs []Request, s Strategy, hitCostMS float64, dropStopwords bool) Result {
	ordered := make([]Request, len(reqs))
	copy(ordered, reqs)
	sort.SliceStable(ordered, func(i, j int) bool {
		if ordered[i].TS == ordered[j].TS {
			return ordered[i].ID < ordered[j].ID
		}
		return ordered[i].TS < ordered[j].TS
	})

	res := Result{Strategy: s.Name(), Total: len(ordered)}
	qualitySum := 0.0
	for _, req := range ordered {
		sig := Derive(req.Prompt, dropStopwords)
		if m, ok := s.Lookup(sig, req.Route); ok {
			served := hitCostMS
			if req.CostMS < served {
				served = req.CostMS
			}
			avoided := req.CostMS - served
			if avoided < 0 {
				avoided = 0
			}
			res.Hits++
			res.ServedMS += served
			res.AvoidedMS += avoided
			qualitySum += m.Quality
		} else {
			s.Insert(req.ID, sig, req.Route)
			res.ServedMS += req.CostMS
		}
	}
	if res.Total > 0 {
		res.HitRate = round4(float64(res.Hits) / float64(res.Total))
	}
	if res.Hits > 0 {
		res.MeanQual = round4(qualitySum / float64(res.Hits))
	}
	res.AvoidedMS = round4(res.AvoidedMS)
	res.ServedMS = round4(res.ServedMS)
	return res
}
